### ISSUE TITLE
PayPalHere: The composite token string was not in the expected format

### ISSUE THREAD

**Issue by cesariel** [paypal/paypal-retail-node#1]
I'm now in a LIVE environment, but I'm receiving this error: 

Data: {"correlationId":"4220c76f7f84e","errorCode":600031,"message":"Missing access token","developerMessage":"You must provide an access token when calling this API. It can be passed as either a header of the form \"Authorization: Bearer <access token>\" or as a query parameter called access_token.","errorType":"oauth/missing_access_token"}

@djMax

**Comment by cesariel** [comment]
This is strange, when https://github.com/paypal/paypal-retail-node/issues/1#issue-108258834 this error doesn't show up... The reader doesn't starts:

Data: {"correlationId":"e3e3e465266b4","preferences":{"ccStmtName":"RODINFINTER"},"fingerprint":"nRqgHy3IP2mav2uf1eA11Q=="}
2015-09-25 00:40:21.471 AmwayBusinessManager[1154:423608] 04:40:20.0899      Main thread | PayPalHere: TXMgr: registerWithCardReaderListener
2015-09-25 00:40:21.474 AmwayBusinessManager[1154:423608] 04:40:20.0899      Main thread | PayPalHere: TXMgr: beginMonitoring with readerMask 0
2015-09-25 00:40:21.474 AmwayBusinessManager[1154:423608] 04:40:20.0900      Main thread | PayPalHere: Dock Port Reader connection status changed from "Idle" to "Idle"
2015-09-25 00:40:21.475 AmwayBusinessManager[1154:423608] 04:40:20.0901      Main thread | PayPalHere: Bluetooth Reader connection status changed from "Idle" to "Idle"
2015-09-25 00:40:21.475 AmwayBusinessManager[1154:423608] 04:40:20.0901      Main thread | PayPalHere: -[PPHCardReaderManager beginMonitoring:] entry
2015-09-25 00:40:21.475 AmwayBusinessManager[1154:423608] 04:40:20.0902      Main thread | PayPalHere: setState: 0
2015-09-25 00:40:21.475 AmwayBusinessManager[1154:423608] 04:40:20.0902      Main thread | PayPalHere: setState: !_ignoreHardwareReaders
2015-09-25 00:40:21.478 AmwayBusinessManager[1154:423608] 04:40:20.0902      Main thread | PayPalHere: SetState: About to emitPaymentEvent for Idle
2015-09-25 00:40:21.478 AmwayBusinessManager[1154:423608] 04:40:20.0902      Main thread | PayPalHere: -[PPHCardReaderManager endTransaction] entry
2015-09-25 00:40:21.478 AmwayBusinessManager[1154:423608] 04:40:20.0902      Main thread | PayPalHere: CardReaderManager::endTransaction
2015-09-25 00:40:21.479 AmwayBusinessManager[1154:423608] 04:40:20.0903      Main thread | PayPalHere: -[PPHTransactionManager cancelPayment] entry
2015-09-25 00:40:21.479 AmwayBusinessManager[1154:423608] 04:40:20.0903      Main thread | PayPalHere: Clearing payment data
2015-09-25 00:40:21.479 AmwayBusinessManager[1154:423608] 04:40:20.0904      Main thread | PayPalHere: -[PPHTransactionManager deActivateReaderForPayments] entry
2015-09-25 00:40:21.480 AmwayBusinessManager[1154:423608] 04:40:20.0904      Main thread | PayPalHere: Flow Type&&Flow Action:DeActivate reader for payments invoked.
2015-09-25 00:40:21.480 AmwayBusinessManager[1154:423608] 04:40:20.0904      Main thread | PayPalHere: setInteger 114 forKey pp.msgIndex
2015-09-25 00:40:21.482 AmwayBusinessManager[1154:423608] 04:40:20.0905      Main thread | PayPalHere: setInteger 115 forKey pp.msgIndex
2015-09-25 00:40:21.482 AmwayBusinessManager[1154:423608] 04:40:20.0908      Main thread | PayPalHere: -[PPHTransactionManager beginPaymentUsingUIWithInvoice:transactionController:] entry
2015-09-25 00:40:21.482 AmwayBusinessManager[1154:423608] 04:40:20.0919      Main thread | PayPalHere: Clearing payment data
2015-09-25 00:40:21.483 AmwayBusinessManager[1154:423608] 04:40:20.0919      Main thread | PayPalHere: setState: 1
2015-09-25 00:40:21.483 AmwayBusinessManager[1154:423608] 04:40:20.0919      Main thread | PayPalHere: setState: !_ignoreHardwareReaders
2015-09-25 00:40:21.483 AmwayBusinessManager[1154:423608] 04:40:20.0919      Main thread | PayPalHere: Watching current invoice.
2015-09-25 00:40:21.483 AmwayBusinessManager[1154:423608] 04:40:20.0920      Main thread | PayPalHere: Credit card flow payment method updated: 0
2015-09-25 00:40:21.485 AmwayBusinessManager[1154:423608] 04:40:20.0920      Main thread | PayPalHere: SDK Flow Controller beginning flow
2015-09-25 00:40:21.486 AmwayBusinessManager[1154:423608] 04:40:20.0978      Main thread | PayPalHere: Flow step became active: PPHSDKWaitForCardFlowStep
2015-09-25 00:40:21.486 AmwayBusinessManager[1154:423608] 04:40:20.0979      Main thread | PayPalHere: EMVSDK Payment Flow: PPHSDKWaitForCardFlowStep&Unique Trace Key: 429BB141-2022-4C1C-9EE8-4BA245431EAA-1154-000001484EA61669&Flow Action:Step Became Active
2015-09-25 00:40:21.486 AmwayBusinessManager[1154:423608] 04:40:20.0979      Main thread | PayPalHere: setInteger 116 forKey pp.msgIndex
2015-09-25 00:40:21.486 AmwayBusinessManager[1154:423608] 04:40:20.0980      Main thread | PayPalHere: [EMV] Waiting for card
2015-09-25 00:40:21.486 AmwayBusinessManager[1154:423608] 04:40:20.0980      Main thread | PayPalHere: Flow Type&Unique Trace Key: 429BB141-2022-4C1C-9EE8-4BA245431EAA-1154-000001484EA61669&Flow Type:Payment
2015-09-25 00:40:21.488 AmwayBusinessManager[1154:423608] 04:40:20.0980      Main thread | PayPalHere: setInteger 117 forKey pp.msgIndex
2015-09-25 00:40:21.488 AmwayBusinessManager[1154:423608] 04:40:21.0010      Main thread | PayPalHere: 
--- _OBSCURO_EMFJSONRequest (POST https://www.paypal.com/webapps/hereapi/merchant/v1/cal)

**Comment by djMax** [comment]
Something is wrong with the token that's being used / generated for live. I assume this is still using the paypal-retail-node server?

**Comment by cesariel** [comment]
Yes. Using the retail node. I get the sdk_token, but when I pass it to the paypal sdk I get this error.

**Comment by cesariel** [comment]
@djMax I change:

1-
[PayPalHereSDK selectEnvironmentWithType:ePPHSDKServiceType_Live];

2- heroku
live secret and live client id

3- toPaypal/live

I don't know what else, I'm doing wrong?

**Comment by cesariel** [comment]
@djMax Tell me if I can help you in something?

**Comment by djMax** [comment]
what's the URL of your copy of paypal-retail-node? (The heroku URL)

**Comment by cesariel** [comment]
Hi @djMax  !!

Thanks!!

The heroku url is:

http://rodinf.herokuapp.com/toPayPal/live

**Comment by djMax** [comment]
Can you show me the code that calls setupWithCompositeTokenString? And what comes back in the completion handler for that call. An error?

**Comment by djMax** [comment]
I've verified that the server is generating proper tokens

**Comment by cesariel** [comment]
@djMax Ok! Let me run it again.

**Comment by cesariel** [comment]
@djMax 

The code calling is:

self.currentUser.pptoken = sdk_token;
        
        [PayPalHereSDK setupWithCompositeTokenString:self.currentUser.pptoken
                               thenCompletionHandler:^(PPHInitResultType status, PPHError *error, PPHMerchantInfo *info) {
                                   
                                   if (!error) {
                                       [self performSegueWithIdentifier: @"showPaymentSegue" sender: self];
                                       NSLog(@"setupWithCompositeTokenString: %ld", (long)status);
                                       NSLog(@"setupWithCompositeTokenString: %@",info);
                                   } else{
                                   
                                       NSLog(@"setupWithCompositeTokenString ERROR: %@", error);
                                   }
                                   
                                   
                                   
                                   
                                   
                               }];


The logs:


2015-09-28 22:23:16.979 AmwayBusinessManager[1793:666787] SDK Token is: live:WyJBMDE1bkt4cjRjQjFKUFBkbDZQRjN0bHo4UkJhcTVIOFlpRm1vLUFmTTdUYTZDMCIsIjI4ODAwIiwiaHR0cDovL3JvZGluZi5oZXJva3VhcHAuY29tL3JlZnJlc2g/JnRva2VuPVRLRWRJJTJGT2hOaE1Ia0VNRUFPUVVmc2RSSWcxdEFMY2gyU25kM1ZaQTl6UWZYRlJTTVNkbUFVN25uNXhBU1NwaFNueXRsdzNsVUJWQTBRT3EzJTJGMzlDdzhZbyUyRjlPYlZVYnVNcU5LMXo4VVlwcFBDUUdqdThncHZHUWJQalZRZWNlOEN5TlV5JTJGMWxsWHBjZll4U01QNXNyMjRYUzVDeThiR3VRa3R5YzIlMkYxZmJaTXNVS2ZDTGFBJTJCaE1nM2xvUUZiWU11UWloNkNXJTJCZ1B5ZFJuMVYwYml3UHBWYzFQbFhxb28zZzRQQWZsem0xa2hDMjlHUHc4eXRxbXpCeEc0Z1haejl2SnFUQSUzRCUzRCJd
2015-09-28 22:23:16.990 AmwayBusinessManager[1793:666787] setupWithCompositeTokenString ERROR: Error Domain=PPHLocal Code=-2007 The composite token string was not in the expected format apiMessage=(null) message=(null) correlationId=(null) @ 2015-09-29 2:23:16 a.m. +0000
2015-09-28 22:23:16.991 AmwayBusinessManager[1793:666787] PayPalHere: The composite token string was not in the expected format
2015-09-28 22:23:16.994 AmwayBusinessManager[1793:666787] Debug: PayPalHere: integerForKey pp.runIndex = 20
2015-09-28 22:23:16.995 AmwayBusinessManager[1793:666787] Debug: PayPalHere: setInteger 21 forKey pp.runIndex
2015-09-28 22:23:16.998 AmwayBusinessManager[1793:666787] Debug: PayPalHere: integerForKey pp.msgIndex = 140
2015-09-28 22:23:16.999 AmwayBusinessManager[1793:666787] Debug: PayPalHere: setInteger 141 forKey pp.msgIndex
2015-09-28 22:23:17.000 AmwayBusinessManager[1793:666787] Debug: PayPalHere: setInteger 142 forKey pp.msgIndex
2015-09-28 22:23:17.634 AmwayBusinessManager[1793:666883] 02:23:16.0981      Main thread | PayPalHere: +[PayPalHereSDK setupWithCompositeTokenString:thenCompletionHandler:] entry
2015-09-28 22:23:17.634 AmwayBusinessManager[1793:666883] 02:23:16.0991      Main thread | PayPalHere: The composite token string was not in the expected format
2015-09-28 22:23:17.635 AmwayBusinessManager[1793:666883] 02:23:16.0995      Main thread | PayPalHere: integerForKey pp.runIndex = 20
2015-09-28 22:23:17.635 AmwayBusinessManager[1793:666883] 02:23:16.0996      Main thread | PayPalHere: setInteger 21 forKey pp.runIndex
2015-09-28 22:23:17.635 AmwayBusinessManager[1793:666883] 02:23:16.0999      Main thread | PayPalHere: integerForKey pp.msgIndex = 140
2015-09-28 22:23:17.636 AmwayBusinessManager[1793:666883] 02:23:16.0999      Main thread | PayPalHere: setInteger 141 forKey pp.msgIndex
2015-09-28 22:23:17.636 AmwayBusinessManager[1793:666883] 02:23:17.0000      Main thread | PayPalHere: setInteger 142 forKey pp.msgIndex

**Comment by djMax** [comment]
(You don't want that live token in github logs, I removed it but I saw it)

**Comment by cesariel** [comment]
Oh!!!!!!

**Comment by cesariel** [comment]
My bad !!

**Comment by djMax** [comment]
Hmmm. The token looks properly formatted but I see the log entry that says the SDK doesn't like it. We'll dig a bit more. Can you get a new token? Or does it always fail?

**Comment by cesariel** [comment]
I can try with a new one! But it always fails at the same point...

**Comment by cesariel** [comment]
Right now, I run it again, with a new token, but the same... 
`2015-09-28 22:51:05.854 AmwayBusinessManager[1829:675029] setupWithCompositeTokenString ERROR: Error Domain=PPHLocal Code=-2007 The composite token string was not in the expected format apiMessage=(null) message=(null) correlationId=(null) @ 2015-09-29 2:51:05 a.m. +0000
2015-09-28 22:51:05.855 AmwayBusinessManager[1829:675029] PayPalHere: The composite token string was not in the expected format`

**Comment by djMax** [comment]
The country of the account you're using doesn't appear to be in a country we support... Terrible error message on our part, which we'll fix, but do you have a US, AU or GB account?

**Comment by djMax** [comment]
the token you posted though, does appear to PARSE correctly, so my big question is are you _sure_ you're passing the token exactly as you had shown in those logs? Something appears to be changing by the time we get it.

**Comment by cesariel** [comment]
I'm from Dominican Republic, the largest bank in the Dominican Republic just made an alliance with Paypal. Now we can draw funds directly from the bank (Banco Popular / Popular Bank). So, I have a Dominican Account :smile: ... Is there a problem?

**Comment by cesariel** [comment]
The only thing that I do, is assign the token to a variable. And pass it to SDK.

**Comment by djMax** [comment]
We don't have card processing services active in DR. Checkin might work (not sure), but CC almost certainly will not. Try NSLog(@"TOKEN >>%@<<", sdk_token) and paste the result (strip out the stuff after live:(first few chars) up until the last few chars)

**Comment by cesariel** [comment]
I pass the sdks_token without assigning to the variable... and now it work !!! Maybe it was because the variable is a Core Data Object!!!

**Comment by cesariel** [comment]
But now, the reader tell me NOT FOUND :disappointed: 

It's because my Country doesn't accept the CC service?

**Comment by djMax** [comment]
yeah, I doubt we activate any of the reader logic for DR accounts.

**Comment by cesariel** [comment]
And what do you recommend ?? Because I'm a developer for the world ... This app, is for global use, not only to DR ?

**Comment by djMax** [comment]
Yeah, for merchants in AU, UK and GB it should work fine (once they've onboarded for card processing). But I guess that means testing is only going to work in sandbox.

**Comment by djMax** [comment]
each country has very unique requirements for both customer signup and card processing, and so we basically take them as there's enough demand to build the necessary infrastructure for direct card processing.

**Comment by cesariel** [comment]
WOAH! Thats a kick to my stomach !!! After all this work and infinite hours of coding !! And asking, and bother you :smile:

**Comment by cesariel** [comment]
@djMax Hi !! 

So I'm thinking about a double solution... For accounts that have Paypal Here and have a US account, and for people like me, that live in another country. 

Whats my way to go? How can I accept payments in my APP without Paypal Here ?? What can I do?

Thank you very much!!

**Comment by djMax** [comment]
Are you able to use the REST APIs in the countries in question? https://developer.paypal.com/docs/api/#create-a-payment

**Comment by cesariel** [comment]
Ok, I will check!!

Meanwhile I have another BIG question! My brother live in USA, and have a Business Account in USA. I configure a new App, in his account, I try to login with his account, but the Paypal Here doesn't work!

@djMax I can't use a USA account to test?

**Comment by djMax** [comment]
Has he "onboarded" for PayPal Here using (for example) the PayPal Here app? If so, it should work. If not, that needs to be done first. How does it fail?

**Comment by cesariel** [comment]
@djMax I'm receiving the Error Code: 67, the Paypal Here App, is working good. 

Data: {"correlationId":"ec0ea6e4c17c6","errorCode":600000,"message":"InvoiceServ: Unknown Error Code: 67","errorType":"service_error"}
2015-10-12 12:44:03.089 AmwayBusinessManager[1214:593501] 16:44:02.0335      Main thread | PayPalHere: setInteger 19 forKey pp.msgIndex
2015-10-12 12:44:03.089 AmwayBusinessManager[1214:593501] 16:44:02.0338      Main thread | PayPalHere: Save Invoice Flow Step failed with error: InvoiceServ: Unknown Error Code: 67
2015-10-12 12:44:03.090 AmwayBusinessManager[1214:593501] 16:44:02.0339      Main thread | PayPalHere: setInteger 20 forKey pp.msgIndex
2015-10-12 12:44:03.092 AmwayBusinessManager[1214:593501] 16:44:02.0350      Main thread | PayPalHere: UNLOCALIZED ERROR - Attempted to display error with full error key: "erservice_error_600000_InvoiceServ:UnknownErrorCode:67_US_title". Please use this key in full or remove any number of its sections separated by "_" to a localized message for this error.
2015-10-12 12:44:03.093 AmwayBusinessManager[1214:593501] 16:44:02.0351      Main thread | PayPalHere: UNLOCALIZED ERROR - Attempted to display error with full error key: "erservice_error_600000_InvoiceServ:UnknownErrorCode:67_US". Please use this key in full or remove any number of its sections separated by "_" to a localized message for this error.
2015-10-12 12:44:03.093 AmwayBusinessManager[1214:593501] 16:44:02.0366      Main thread | PayPalHere: [SDK Alert] {
  "title" : "PPH_EMVSDK_Error",
  "message" : "The operation couldn’t be completed. (service_error error 600000.) - InvoiceServ: Unknown Error Code: 67",
  "buttons" : [
    "PPH_Label_Retry",
    "PPH_EMVSDK_CancelLabel"
  ]
}

**Comment by djMax** [comment]
Usually this means the app doesn't have the proper scopes. Can you send the invoice JSON? Is it missing currency code?

**Comment by djMax** [comment]
That transaction was in live? I don't find that correlationId which is odd.

**Comment by cesariel** [comment]
Was live

**Comment by cesariel** [comment]
Yes! Was the currency. I had "DOP"... I change to "USD"... But now, I'm receiving another error...

![img_2554](https://cloud.githubusercontent.com/assets/13699157/10433816/0b1a7070-70e3-11e5-8308-e3cef2d1873f.PNG)

Data: {"correlationId":"a864cef3dabf5","errorCode":580031,"message":"Please try another card","developerMessage":"This credit card cannot be processed by PayPal","errorType":"pay/bad_card"}

@djMax

**Comment by cesariel** [comment]
I try, with 2 different type of Credit Card:

Amex
Visa

But the same Error

**Comment by djMax** [comment]
I can't tell you definitively why, but basically our risk engine declined the transaction. Obviously you're trying to test, and somehow our systems have figured out this isn't a "normal" transaction and blocked it. Apologies, though I think everything is working from your app.

**Comment by cesariel** [comment]
Do you think is maybe because I use the DEBUG SDK ? @djMax

**Comment by javagc** [comment]
@cesariel  have you found solution? i have same issue. Thanks in advance
